### PR TITLE
Atmega2560p CODE for HAL

### PR DESCRIPTION
**Pull request by dkjha20**
Make sure these boxes are checked! 📦✅

- [x] You ran `toolchain.sh` to install correct rust `nightly-2021-01-07`
- [x] You have the latest version of `rustfmt` installed
```bash
$ rustup component add rustfmt
```
- [x] You ran `cargo fmt` on the code base before submitting
- [x] You reference which issue is being closed in the PR text
- [x] You added proper comments to the code
- [x] Your code compiled correctly for the respective board
- [x] You did not edit files outside your issue

✨✨ 😄 Great work, continue contributing to the project! 😄 ✨✨

### CODE REVIEW

File: examples/atmega2560p/blink/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#![no_std]
+#![no_main]
+#![deny(warnings)]
+
+/// Crates included which are to be used for the Blinking LED example.
+use rustduino::atmega2560p::hal::pin::Pins;
+use rustduino::atmega2560p::hal::watchdog::WatchDog;
+
+#[no_mangle]
+pub extern "C" fn main() {
```
Comment (Mshivam2409) [inline examples/atmega2560p/blink/src/main.rs]: It's not a extern "C" function.

File: examples/atmega2560p/blink/target/CACHEDIR.TAG [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Signature: 8a477f597d28d172789f06886806bc55
```
Comment (Mshivam2409) [inline examples/atmega2560p/blink/target/CACHEDIR.TAG]: Target folder should not be included, remove everything from that

File: src/atmega2560p/hal/interrupts.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+//
+//     You should have received a copy of the GNU Affero General Public License
+//     along with this program.  If not, see <https://www.gnu.org/licenses/>
+
+//! Global interrupts configured in the ATMEGA2560P chip is controlled here.
+//! Section 7.4 of the manual
+//! https://ww1.microchip.com/downloads/en/devicedoc/atmel-2549-8-bit-avr-microcontroller-atmega640-1280-1281-2560-2561_datasheet.pdf
+
+/// Crates required in the code for reading and writing to registers.
+use core;
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/interrupts.rs]: Don't import the whole crate.

File: src/atmega2560p/hal/pin.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+//     GNU Affero General Public License for more details.
+//
+//     You should have received a copy of the GNU Affero General Public License
+//     along with this program.  If not, see <https://www.gnu.org/licenses/>
+
+/// Include the required crates for the code.
+use crate::atmega2560p::hal::port::*;
+
+/// All pins inside a single struct.
+pub struct Pins {
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/pin.rs:25]: Give a more descriptive comment.

File: src/atmega2560p/hal/port.rs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    K,
+    L,
+}
+
+/// Each Port pin consists of 3 registers : PINx, DDRx, PORTx .
+/// These registors control the ports.
+/// ```DDRx:  Data direction register```
+///     The DDRx Register controls the direction of a particular pin.
+/// ```PORTx: Data register```
+///     PORTxn is written logic one when the pin is configured as an input pin.
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/port.rs]: We don't need register manipulation in comments.

File: src/atmega2560p/hal/port.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    port: u8,
+}
+
+/// The structure Pin contains the address of the port to which the pin belongs and the pin's number.
+pub struct Pin {
+    port: *mut Port,
+    pin: usize,
+}
+
+/// Type ```IOMode```
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/port.rs]: Why triple ticks?

File: src/atmega2560p/hal/port.rs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+impl Pin {
+    ///Return a pin for the given port name and pin number.
+    pub fn new(port: PortName, pin: usize) -> Option<Pin> {
+        Port::new(port).pin(pin)
+    }
+
+    /// Change pin mode to input or output by changing the DDr register.
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/port.rs:127]: Better comment.

File: src/atmega2560p/hal/port.rs [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            IOMode::Input => 0x0,
+            IOMode::Output => 0x1 << self.pin,
+        };
+
+        // Write the value to DDxn register.
+        unsafe { write_volatile(&mut (*self.port).ddr, ddr_val) }
+    }
+
+    /// Toggles the value of PORTxn by writing one to PINxn ,independent of the value of DDRxn.
+    pub fn toggle(&mut self) {
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/port.rs:146]: Again we don't need what happens inside, I need to know what toggle does for me.

File: src/atmega2560p/hal/power.rs [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    TIMER5,
+    TIMER4,
+    TIMER3,
+    USART3,
+    USART2,
+    USART1,
+}
+
+/// Contains registers to control the functioning of clocks in the chip -
+/// PRR0 : Contains 7 Writable bits which will be used to control some of the clocks modes.
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/power.rs]: I'm seeing register comments everywhere.

File: src/atmega2560p/hal/power.rs [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+            },
+        }
+    }
+
+    /// This is the function for enabling the clock system of your choice.
+    /// Here we would create a new element of the structure power
+    /// and it would be used to control various clock gating features of the
+    /// chip ATMEGA2560P.
+    /// All the clock features are implemented in this function using match cases.
+    pub fn enable_clocks(&mut self, mode: Options) {
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/power.rs:172]: Please let the user know, what exactly is options i.e. what he has to pass in as second argument.

File: src/atmega2560p/hal/watchdog.rs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+use core;
+
+/// use interrupts to enable/disable global interrupts
+/// section 28.6 from datasheet for atmega2560p
+/// prior to disabling watchdog, all interrupts must be disabled
+/// when WDE and WDIE bits of WDTCSR register  sets to 0, watchdog disables
+/// WDRF bit of MCUSR register can overwrite WDE ,so,WDRF must be cleared before
+///
+
+/// WatchDog struct Contains various registers to control the functioning of registers Watchdog.
```
Comment (Mshivam2409) [inline src/atmega2560p/hal/watchdog.rs]: Again, registers

**Review by Mshivam2409 — CHANGES_REQUESTED** [review]

File: src/atmega2560p/hal/watchdog.rs (hunk 11 above)
Comment (dkjha20) [reply src/atmega2560p/hal/watchdog.rs:32]: Adjusted

File: src/atmega2560p/hal/power.rs (hunk 10 above)
Comment (dkjha20) [reply src/atmega2560p/hal/power.rs:166]: Adjusted, BTW it was mentioned as another comment in the starting of the code also. I have briefed it here again.

File: src/atmega2560p/hal/power.rs (hunk 9 above)
Comment (dkjha20) [reply src/atmega2560p/hal/power.rs:57]: Adjusted, sorry for bothering.

File: src/atmega2560p/hal/port.rs (hunk 5 above)
Comment (dkjha20) [reply src/atmega2560p/hal/port.rs:48]: fixed

File: src/atmega2560p/hal/interrupts.rs (hunk 3 above)
Comment (dkjha20) [reply src/atmega2560p/hal/interrupts.rs:22]: adjusted

File: src/atmega2560p/hal/port.rs (hunk 6 above)
Comment (dkjha20) [reply src/atmega2560p/hal/port.rs:65]: I was correlating it with MARKDOWN. Sorry for that, adjusted everywhere.

File: src/atmega2560p/hal/port.rs (hunk 7 above)
Comment (dkjha20) [reply src/atmega2560p/hal/port.rs:127]: Thanks :)

File: src/atmega2560p/hal/port.rs (hunk 8 above)
Comment (dkjha20) [reply src/atmega2560p/hal/port.rs:144]: fixed

File: src/atmega2560p/hal/pin.rs (hunk 4 above)
Comment (dkjha20) [reply src/atmega2560p/hal/pin.rs:21]: done

File: examples/atmega2560p/blink/src/main.rs (hunk 1 above)
Comment (dkjha20) [reply examples/atmega2560p/blink/src/main.rs:26]: adjusted

File: examples/atmega2560p/blink/target/CACHEDIR.TAG (hunk 2 above)
Comment (dkjha20) [reply examples/atmega2560p/blink/target/CACHEDIR.TAG:1]: deleted from the repository

**Review by Mshivam2409 — APPROVED** [review]

**Pull request merged**